packages/contracts/contracts/examples/UnirepVoting.sol

```diff
@@ -0,0 +1,148 @@
+        uint256[8] calldata proof
+    ) public {
+        unirep.userSignUp(publicSignals, proof);
+    }
+
+    function joinProject(
+        uint256 projectID,
+        uint256[] memory publicSignals,
+        uint256[8] memory proof
+    ) public {
+        require(projectID < numTeams, 'Voteathon: invalid project iD');
+        EpochKeyVerifierHelper.EpochKeySignals memory signals = epochKeyHelper.verifyAndCheck(publicSignals, proof);
+
+        require(signals.revealNonce == true, 'Voteathon: should reveal nonce');
+        require(signals.nonce == 0, 'Voteathon: invalid nonce');
+        require(counts[projectID] < 10, 'Voteathon: maximum participants in a project');
+        participants[projectID].push(signals.epochKey);
+        // give user data if there is attestation before
+        uint48 epoch = unirep.attesterCurrentEpoch(uint160(address(this)));
+        require(epoch == 0, 'Voteathon: not join epoch');
+        uint256[] memory data = projectData[projectID];
+        for (uint256 i = 0; i < data.length; i++) {
+            unirep.attest(signals.epochKey, epoch, i, data[i]);
+        }
+        counts[projectID] += 1;
+    }
+
+    function vote(
+        uint256 projectID,
+        Option option,
+        uint256[] calldata publicSignals,
+        uint256[8] calldata proof
+    ) public {
+        require(projectID < numTeams, "projectID out of range");
+
+        EpochKeyVerifierHelper.EpochKeySignals memory signals = epochKeyHelper.verifyAndCheck(publicSignals, proof);
+
+        require(signals.revealNonce == true, "reveal nonce wrong");
+        require(signals.nonce == 1, "nonce wrong");
+
+        projectData[projectID][uint(option)] += 1;
+        
+        voted[signals.epochKey] += 1;
+        if (option == Option.UP) scores[projectID] += 1;
+        else if (option == Option.DOWN) scores[projectID] -= 1;
+
+        uint[] memory members = participants[projectID];
+        uint48 epoch = unirep.attesterCurrentEpoch(uint160(address(this)));
+        require(epoch == 0, 'Voteathon: not voting epoch');
+        for (uint256 i = 0; i < members.length; i++) {
+            unirep.attest(members[i], epoch, uint(option), 1);
+        }
+        
+        uint160 attesterId = uint160(address(this));
+        unirep.attesterStateTreeRootExists(attesterId, epoch, signals.stateTreeRoot);
+    }
+    
+    function claimPrize(
+        address receiver,
+        uint256[] calldata publicSignals,
+        uint256[8] calldata proof
+    ) public {
+        uint160 attesterId = uint160(address(this));
+        require(unirep.attesterCurrentEpoch(attesterId) > 0);
+        ReputationVerifierHelper.ReputationSignals memory signals = helper.verifyAndCheck(publicSignals, proof);
+        uint48 epoch = 0; // the voting epoch
+        uint256 stateTreeRoot = signals.stateTreeRoot;
+        unirep.attesterStateTreeRootExists(attesterId, epoch, stateTreeRoot);
+        require(!claimed[publicSignals[1]], 'Already claimed');
```

Comment:
we can also verify the `revealNonce` and `nonce` in this function.

```solidity
        require(signals.revealNonce == true, "reveal nonce wrong");
        require(signals.nonce == 1, "nonce wrong");
```
to make sure the user doesn't use two epoch keys to receive prize
and there should be a `signals.epochKey` in [`ReputationSignals`](https://developer.unirep.io/docs/contracts-api/verifiers/reputation-verifier-helper) you can use instead of using `publicSignals[1]`